File: 1008_construct_binary_search_tree_from_preorder_traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from binarytree import TreeNode, buildtree, leetcode_tree_repr, is_same_tree

class Solution(object):
    def bstFromPreorder(self, preorder):
        """
        :type preorder: List[int]
        :rtype: Optional[TreeNode]
        """
        def tree_insert(root, val):
            new = TreeNode(val)
            if root is None:
                return new
            parent = None
            current = root
            while current is not None:
                parent = current
                if current.val > new.val:
                    current = current.left
                elif current.val < new.val:
                    current = current.right
                else:
                    return root
            if parent.val > new.val:
                parent.left = new
            else:
                parent.right = new
            return root
        root = None
        for i in range(0, len(preorder)):
            root = tree_insert(root, preorder[i])
        return root

vectors = [
        [8,5,1,7,10,12], [8,5,10,1,7,None,12],
        [1,3], [1,None,3]
        ]

for i in range(0, len(vectors), 2):
    nums = vectors[i]
    b = vectors[i+1]
    print(nums)
    expected = buildtree(b)
    returned = Solution().bstFromPreorder(nums)
    a = leetcode_tree_repr(returned)
    assert is_same_tree(expected, returned), f'for {nums} expected {b}, but returned {a}!'

File: C/1008_construct_binary_search_tree_from_preorder_traversal.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

static struct TreeNode* allocTreeNode(int val, struct TreeNode *left, struct TreeNode *right) {
    struct TreeNode *node;

    node = malloc(sizeof(*node));
    if (node) {
        node->val = val;
        node->left = left;
        node->right = right;
    }
    return node;
}
#define ALLOC_TREE_NODE(val) allocTreeNode((val), NULL, NULL)

static struct TreeNode* bstInsert(struct TreeNode *root, int val) {
    struct TreeNode *new = ALLOC_TREE_NODE(val);
    struct TreeNode *parent, *current;

    if (!root)
        return new;

    parent = NULL;
    current = root;
    while (current) {
        parent = current;
        if (current->val > new->val) {
            current = current->left;
        } else if (current->val < new->val) {
            current = current->right;
        } else {
            /* should not be here, but... */
            return root;
        }
    }
    if (parent->val > new->val) {
        parent->left = new;
    } else {
        parent->right = new;
    }
    return root;
}

struct TreeNode* bstFromPreorder(int* preorder, int preorderSize) {
    struct TreeNode *root = NULL;
    int i;

    if (preorder && preorderSize) {
        for (i = 0; i < preorderSize; i++) {
            root = bstInsert(root, *preorder);
            preorder++;
        }
    }
    return root;
}
